Hack-A-Sat 2021/Presents From Marco/Groundead/README.md

```diff
@@ -0,0 +1,154 @@
+
+    >8 0 0 1 0 1 7
+    8 0 0 1 0 1 7
+    
+    Packet Version Number: 00000000
+    Packet Type: 00000000
+    Secondary Header Flag: 00000001
+    Application Process Identifier: 00000000
+    Sequence Flags: 00000000
+    Packet Sequence Count or Packet Name: 00000001
+    Packet Data Length: 00000002
+
+
+    That sequence of hex characters did not work. Try again.
+
+The above are debug messages output by this function. We also see the above message is outout if this function does not successfully validate the packet header, or after all supplied packets have been processed.
+
+## Exploit Strategy
+
+Given that `getSatellitePacketBytes` forces the 7th octet to be '7' and that this is the 'mode' option that we need to be '8' we need to find a way of overcoming that. It is noted that the PacketDataLength is included in the Packet Primary Header so we investigate whether we could send two concatenated packets in one line of input with only the first one having the 'mode 7' enforced.
+
+We know that we can take advantage of the timing to send multiple lines to the queue, how can we spoof the contents to trigger processing of an additional packet that bypasses the `mode` being forced to 7?
+
+Further analysis of the code reveals:
+
+1. Only the first 32 bytes of the input buffer are cleared to zero's for each input line by `getSatellitePacketBytes`.
+2. The payload data scanned for 5 occurrences of "1acffc1d" by `processSatellitePacketBytes`.
+
+We determined that a pair of messages could be sent, with the first message satisfying the expected payload data and the second one declaring a shorter packet length than the supplied data such that the reamining octetswould be treated as a further packet.
+
+## Final Exploit
+
+Some tweaks were made to the exploit candidate to ensure that all of the checks on payload length, payload data contents and mode were satisfied. The final exploit input is as follows:
+
+    >8 0 0 1 0 1a 7 1a cf fc 1d 1a cf fc 1d 1a cf fc 1d 1a cf fc 1d 1a cf fc 1d 0 0 0 0 0 8
+    8 0 0 1 0 1a 7 1a cf fc 1d 1a cf fc 1d 1a cf fc 1d 1a cf fc 1d 1a cf fc 8 0 0 1 0 0 8
+    
+    
+    Packet Version Number: 00000000
+    Packet Type: 00000000
+    Secondary Header Flag: 00000001
+    Application Process Identifier: 00000000
+    Sequence Flags: 00000000
+    Packet Sequence Count or Packet Name: 00000001
+    Packet Data Length: 0000001b
+    
+    
+    Handling Test Telemetry
+    
+    
+    Packet Version Number: 00000000
+    Packet Type: 00000000
+    Secondary Header Flag: 00000001
+    Application Process Identifier: 00000000
+    Sequence Flags: 00000000
+    Packet Sequence Count or Packet Name: 00000001
+    Packet Data Length: 00000001
+    
+    
+    EMERGENCY_MODE: THE SPACECRAFT IS IN EMERGENCY_MODE
+    You made it!
+    Here's your flag:
+
+Thank you for providing this challenge. It was great to practice our skills reverse engineering C++ code.
```

Comment:
Was there a particular clue that indicated the original source was in C++? If so that might be interesting to point out.